File: common/src/main/java/net/johnpgr/craftingtableiifabric/block/CraftingTableIIBlock.java
package net.johnpgr.craftingtableiifabric.block;

import com.mojang.serialization.MapCodec;
import net.johnpgr.craftingtableiifabric.CraftingTableII;
import net.johnpgr.craftingtableiifabric.entity.CraftingTableIIEntity;
import net.johnpgr.craftingtableiifabric.platform.Services;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.BaseEntityBlock;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.RenderShape;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CraftingTableIIBlock extends BaseEntityBlock {
    public static final VoxelShape SHAPE = Block.box(1.0, 0.0, 2.0, 15.0, 16.0, 15.0);

    public CraftingTableIIBlock(Properties properties) {
        super(properties);
        registerDefaultState(stateDefinition.any().setValue(BlockStateProperties.HORIZONTAL_FACING, Direction.NORTH));
    }

    public CraftingTableIIBlock() {
        this(Properties.ofFullCopy(Blocks.CRAFTING_TABLE));
    }

    @Override @NotNull
    protected MapCodec<? extends BaseEntityBlock> codec() {
        return simpleCodec(CraftingTableIIBlock::new);
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(BlockStateProperties.HORIZONTAL_FACING);
    }

    @Override
    public BlockState getStateForPlacement(net.minecraft.world.item.context.BlockPlaceContext context) {
        Direction facing = context.getHorizontalDirection().getCounterClockWise();
        if (facing == Direction.NORTH || facing == Direction.SOUTH) {
            facing = facing.getOpposite();
        }
        return defaultBlockState().setValue(BlockStateProperties.HORIZONTAL_FACING, facing);
    }

    @Override @NotNull
    public BlockState rotate(BlockState state, Rotation rotation) {
        return state.setValue(BlockStateProperties.HORIZONTAL_FACING, rotation.rotate(state.getValue(BlockStateProperties.HORIZONTAL_FACING)));
    }

    @Override @NotNull
    public BlockState mirror(BlockState state, Mirror mirror) {
        return state.rotate(mirror.getRotation(state.getValue(BlockStateProperties.HORIZONTAL_FACING)));
    }

    @Override @NotNull
    public RenderShape getRenderShape(BlockState state) {
        return RenderShape.ENTITYBLOCK_ANIMATED;
    }

    @Override
    public boolean propagatesSkylightDown(BlockState state, BlockGetter level, BlockPos pos) {
        return true;
    }

    @Override @NotNull
    public VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
        return SHAPE;
    }

    @Override @NotNull
    public VoxelShape getCollisionShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
        return SHAPE;
    }

    @Nullable
    @Override
    public BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return new CraftingTableIIEntity(pos, state);
    }

    @Nullable
    @Override
    public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level level, BlockState state, BlockEntityType<T> type) {
        return createTickerHelper(type, CraftingTableII.ENTITY_TYPE, CraftingTableIIEntity::tick);
    }

    @Override @NotNull
    protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hit) {
        if (level.isClientSide) {
            return InteractionResult.SUCCESS;
        }

        if (player instanceof ServerPlayer serverPlayer) {
            Services.PLATFORM.openMenu(serverPlayer, pos);
        }

        return InteractionResult.CONSUME;
    }

    @Override
    public boolean hasAnalogOutputSignal(BlockState state) {
        return false;
    }
}

File: src/main/kotlin/net/johnpgr/craftingtableiifabric/block/CraftingTableIIBlock.kt
package net.johnpgr.craftingtableiifabric.block

import com.mojang.serialization.MapCodec
import net.fabricmc.fabric.api.itemgroup.v1.ItemGroupEvents
import net.johnpgr.craftingtableiifabric.CraftingTableIIMod
import net.johnpgr.craftingtableiifabric.block.entity.CraftingTableIIBlockEntity
import net.johnpgr.craftingtableiifabric.screen.CraftingTableIIScreenHandler
import net.johnpgr.craftingtableiifabric.util.BlockScreenHandlerFactory
import net.minecraft.block.Block
import net.minecraft.block.BlockRenderType
import net.minecraft.block.BlockState
import net.minecraft.block.BlockWithEntity
import net.minecraft.block.Blocks
import net.minecraft.block.ShapeContext
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityTicker
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.item.ItemGroups
import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.Items
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.registry.RegistryKey
import net.minecraft.registry.RegistryKeys
import net.minecraft.state.StateManager
import net.minecraft.state.property.Properties
import net.minecraft.util.ActionResult
import net.minecraft.util.BlockMirror
import net.minecraft.util.BlockRotation
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.world.BlockView
import net.minecraft.world.World

class CraftingTableIIBlock : BlockWithEntity(
    Settings.copy(Blocks.CRAFTING_TABLE).registryKey(REGISTRY_KEY)
) {
    companion object {
        val ID = CraftingTableIIMod.id("crafting_table_ii")
        val REGISTRY_KEY: RegistryKey<Block> =
            RegistryKey.of(RegistryKeys.BLOCK, ID)
        val ITEM_REGISTRY_KEY: RegistryKey<Item> =
            RegistryKey.of(RegistryKeys.ITEM, ID)
        val CODEC: MapCodec<CraftingTableIIBlock> =
            createCodec { CraftingTableIIBlock() }

        fun register() {
            Registry.register(Registries.BLOCK, ID, CraftingTableIIMod.BLOCK)

            Registry.register(
                Registries.ITEM, ITEM_REGISTRY_KEY, BlockItem(
                    CraftingTableIIMod.BLOCK,
                    Item.Settings().useBlockPrefixedTranslationKey()
                        .registryKey(ITEM_REGISTRY_KEY)
                )
            )

            ItemGroupEvents.modifyEntriesEvent(ItemGroups.FUNCTIONAL)
                .register { content ->
                    content.addAfter(
                        Items.CRAFTING_TABLE, CraftingTableIIMod.BLOCK
                    )
                }
        }
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(Properties.HORIZONTAL_FACING)
    }

    // FIXME: Maybe this is an unnecessary hack
    override fun getPlacementState(ctx: ItemPlacementContext): BlockState {
        var direction = ctx.horizontalPlayerFacing.rotateYCounterclockwise()
        if (direction == Direction.NORTH || direction == Direction.SOUTH) {
            direction = direction.opposite
        }

        return defaultState.with(Properties.HORIZONTAL_FACING, direction)
    }

    override fun rotate(
        state: BlockState,
        rotation: BlockRotation
    ): BlockState {
        return state.with(
            Properties.HORIZONTAL_FACING,
            rotation.rotate(state[Properties.HORIZONTAL_FACING])
        )
    }

    override fun mirror(
        state: BlockState,
        mirror: BlockMirror
    ): BlockState {
        return state.rotate(mirror.getRotation(state[Properties.HORIZONTAL_FACING]))
    }

    override fun getRenderType(state: BlockState): BlockRenderType {
        return BlockRenderType.MODEL
    }

    override fun hasSidedTransparency(state: BlockState): Boolean {
        return true
    }

    override fun getCollisionShape(
        state: BlockState,
        view: BlockView,
        pos: BlockPos,
        context: ShapeContext
    ): VoxelShape {
        return createCuboidShape(1.0, 0.0, 2.0, 15.0, 16.0, 15.0)
    }

    override fun getOutlineShape(
        state: BlockState,
        world: BlockView,
        pos: BlockPos,
        context: ShapeContext
    ): VoxelShape {
        return createCuboidShape(1.0, 0.0, 2.0, 15.0, 16.0, 15.0)
    }

    override fun createBlockEntity(
        pos: BlockPos,
        state: BlockState
    ): BlockEntity {
        return CraftingTableIIBlockEntity(pos, state)
    }

    override fun getCodec(): MapCodec<out BlockWithEntity> {
        return CODEC
    }

    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hit: BlockHitResult
    ): ActionResult {
        if (world.isClient) {
            return ActionResult.SUCCESS
        }

        player.openHandledScreen(
            BlockScreenHandlerFactory(this, pos, ::CraftingTableIIScreenHandler)
        )

        return ActionResult.CONSUME
    }

    override fun <T : BlockEntity> getTicker(
        world: World,
        state: BlockState,
        type: BlockEntityType<T>
    ): BlockEntityTicker<T> {
        return validateTicker(
            type,
            CraftingTableIIMod.ENTITY_TYPE
        ) { world1, pos, state1, entity -> entity.tick() }!!
    }

    override fun hasComparatorOutput(state: BlockState): Boolean {
        return false
    }
}